prow/tide/codereview.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CodeReviewCommon struct {
+	// NameWithOwner is from graphql.NameWithOwner, <org>/<repo>
+	NameWithOwner string
+	// The number of PR
+	Number int
+	Org    string
+	Repo   string
+	// BaseRefPrefix gets prefix of ref, such as /refs/head, /refs/tags
+	BaseRefPrefix string
+	BaseRefName   string
+	HeadRefName   string
+	HeadRefOID    string
+
+	Title       string
+	Body        string
+	AuthorLogin string
+	// Labels gets labels on the PR.
+	// TODO(chaodaiG): labels might mean something different on gerrit, consider
+	// how to name this nicely.
```

Comment:
IME when trying to different underlying apis away they will always end up having subtle but important differences, even if they look very similiar. So while it seems ok to me to to have a common `data` type, the code that actually interprets this should be seperated or it will end up being riddled with `if isGithub {} else {}` sooner or later. This doesn't only apply to the obviously provider-dependent parts like say `search` but also to something like `filterSubpool` (i.E. there should be one for GitHub, one for Gerrit. They might end up sharing code but should be kept distinct).

This in turn also means that we can be more explicit here and don't need to to try equalize GitHub and Gerrit. I don't know Gerrit very well, but in doubt we could add something like `GerritLabels`.

Writing this out, I am starting to wonder if it would be useful to use substructs, i.E. comprise this out of a `CommonCodeReview`, `GitHubCodeReview`, `GerritCodeReview` struct to more clearly denoate what concepts apply where. If we embedd the substructs and use `jsoin:",inline`, this won't affect serialization and thus won't break Deck.

Reply:
Or maybe the `common` part doesn't need to be a substruct, but moving the fields that are provider-specific into their own substruct would be nice and make it more clear what fields can be assumed to be set when.

Reply:
This is a good point. Had an offline discussion with Cole and we both liked this idea, with concern of potential collisions between embedded structs. So ran an experiment https://go.dev/play/p/vp502rgfVpN, TLDR is that if there is any shared field between two embedded structs, yaml unmarshal just discards the field instead of assigning the value to both, so there is potential risk of certain fields getting lost during repeated marshal/unmarshal. The safest way to do so is probably substruct like:
```
type CommonCodeReview struct {
  // Common fields consumed by Deck
  Org string
  // Provider specific
  GitHub *GitHubCodeReview
  Gerrit *GerritCodeReview
}
```

Also agree with the concern over `if isGithub {} else if isGerrit {}`, I think we can try to cut a line that if we ever need to do this then the field should not be a common field. For instance `Labels` is not a common field.